phfs.c

```diff
@@ -13,61 +13,388 @@
+#include "phfs.h"
+#include "devs.h"
 #include "hal.h"
+
+#include "phoenixd.h"
 #include "plostd.h"
-#include "phfs.h"
+#include "errors.h"
+
+
+#define SIZE_PHFS_HANDLERS    8
+#define SIZE_PHFS_FILES       10
+
+
+enum { phfs_prot_raw = 0, phfs_prot_phoenixd };
 
-/* Configuration file for specific platform */
-#include "config.h"
 
+typedef struct {
+	char alias[8];
+	unsigned int major;
+	unsigned int minor;
+	unsigned int prot;
 
-phfs_handler_t phfs_handlers[PDN_NB];
+	dev_handler_t *h;
+} phfs_handler_t;
 
 
-handle_t phfs_open(u16 pdn, const char *name, u32 flags)
+typedef struct {
+	char alias[32];
+	addr_t addr;
+	size_t size;
+} phfs_file_t;
+
+
+struct {
+	phfs_handler_t handlers[SIZE_PHFS_HANDLERS];
+	unsigned int hCnt;
+
+	phfs_file_t files[SIZE_PHFS_FILES];
+	unsigned int fCnt;
+} phfs_common;
+
+
+static const char *phfs_getProtName(unsigned int id)
 {
-	handle_t handle;
+	switch (id) {
+	case phfs_prot_raw:
+		return "raw";
+
+	case phfs_prot_phoenixd:
+		return "phoenixd";
 
-	if (pdn >= PDN_NB) {
-		handle.h = ERR_ARG;
-		return handle;
+	default:
+		return "";
 	}
+}
+
 
-	return phfs_handlers[pdn].open(phfs_handlers[pdn].dn, name, flags);
+static int phfs_setProt(phfs_handler_t *ph, const char *prot)
+{
+	size_t sz = plostd_strlen(prot);
+
+	if (prot == NULL || (sz == 3 && plostd_strncmp(prot, "raw", 3) == 0))
+		ph->prot = phfs_prot_raw;
+	else if (sz == 8 && plostd_strncmp(prot, "phoenixd", 8) == 0)
+		ph->prot = phfs_prot_phoenixd;
+	else
+		return ERR_ARG;
+
+	return ERR_NONE;
+}
+
+
+static int phfs_getHandlerId(const char *alias)
+{
+	int i;
+	phfs_handler_t *ph;
+	size_t sz, aliasSz = plostd_strlen(alias);
+
+	for (i = 0; i < phfs_common.hCnt; ++i) {
+		ph = &phfs_common.handlers[i];
+		sz = plostd_strlen(ph->alias);
+		if ((sz == aliasSz) && (plostd_strncmp(alias, ph->alias, sz) == 0))
+			return i;
+	}
+
+	return ERR_ARG;
 }
 
 
-s32 phfs_read(u16 pdn, handle_t handle, addr_t *pos, u8 *buff, u32 len)
+int phfs_regDev(const char *alias, unsigned int major, unsigned int minor, const char *prot)
 {
-	if (pdn >= PDN_NB)
+	size_t sz;
+	dev_handler_t *dh;
+	phfs_handler_t *ph;
+
+	if (alias == NULL)
+		return ERR_ARG;
+
+	if (phfs_common.hCnt >= SIZE_PHFS_HANDLERS) {
+		plostd_printf(ATTR_ERROR, "\nToo many devices");
+		return ERR_ARG;
+	}
+
+	if (devs_getHandler(major, minor, &dh) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%d.%d - device does not exist", major, minor);
 		return ERR_ARG;
+	}
+
+	/* Check if alias is already in use */
+	if (phfs_getHandlerId(alias) >= 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - alias is already in use", alias);
+		return ERR_ARG;
+	}
+
+	ph = &phfs_common.handlers[phfs_common.hCnt];
+	if (phfs_setProt(ph, prot) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - wrong protocol name\n\t use: \"%s\", \"%s\"", prot,
+		              phfs_getProtName(phfs_prot_raw), phfs_getProtName(phfs_prot_phoenixd));
+		return ERR_ARG;
+	}
+
+	sz = plostd_strlen(alias);
+	sz = (sz < 7) ? (sz + 1) : 7;
```

Comment:
Instead of magic number `7` use `sizeof(ph->alias) - 1`.
Maybe add `plostd_strncpy()` to such copies?

Reply:
Yes, we can add `plostd_strncpy()` but not in this PR.